File: packages/sdk/src/core/utils/retry.ts
import { HttpError, type Result } from '../types';

const RETRY_DELAYS = [100, 500, 2000] as const;

export async function withRetry<T>(fn: () => Promise<T>): Promise<Result<T>> {
  let lastError: Error | undefined;

  for (let attempt = 0; attempt < RETRY_DELAYS.length + 1; attempt++) {
    try {
      const result = await fn();
      return { success: true, data: result };
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));

      if (lastError instanceof HttpError && !lastError.isRetryable()) {
        break;
      }

      const delay = RETRY_DELAYS[attempt];
      if (delay === undefined) {
        break;
      }

      await new Promise<void>((resolve) => setTimeout(resolve, delay));
    }
  }

  return {
    success: false,
    error: lastError ?? new Error('Unknown error occurred'),
  };
}

File: packages/swift/Sources/PhaseAnalytics/Core/Utils/Retry.swift
import Foundation

internal struct Retry {
    private static let retryDelays: [TimeInterval] = [0.1, 0.5, 2.0]

    static func withRetry<T: Sendable>(
        _ operation: @Sendable @escaping () async throws -> T
    ) async -> Result<T, PhaseError> {
        var lastError: Error?

        for (attempt, delay) in ([0.0] + retryDelays).enumerated() {
            if Task.isCancelled {
                return .failure(.networkError("Request cancelled"))
            }

            if attempt > 0 {
                do {
                    try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                } catch is CancellationError {
                    return .failure(.networkError("Request cancelled"))
                } catch {
                    // Ignore other sleep errors
                }
            }

            do {
                let result = try await operation()
                return .success(result)
            } catch let error as HTTPError {
                lastError = error
                if !error.isRetryable() {
                    break
                }
            } catch is CancellationError {
                return .failure(.networkError("Request cancelled"))
            } catch {
                lastError = error
                break
            }
        }

        let errorMessage = lastError?.localizedDescription ?? "Unknown network error"
        return .failure(.networkError(errorMessage))
    }
}
